omod/src/test/java/org/openmrs/web/controller/ForgotPasswordFormControllerTest.java

```diff
@@ -0,0 +1,281 @@
+			request.addParameter("secretAnswer", "invalid secret answer");
+			
+			controller.handleRequest(request, new MockHttpServletResponse());
+		}
+		
+		// those were the first five, now the sixth request (with a valid user) should fail
+		MockHttpServletRequest request = new MockHttpServletRequest();
+		request.setMethod("POST");
+		
+		request.addParameter("uname", "validuser");
+		request.addParameter("secretAnswer", "valid secret answer");
+		
+		controller.handleRequest(request, new MockHttpServletResponse());
+		
+		Assert.assertFalse(Context.isAuthenticated());
+	}
+	
+	/**
+	 * If a user enters 4 username requests, the 5th one should reset the lockout and they should be
+	 * allowed 5 attempts at the secret answer
+	 *
+	 * @throws Exception
+	 */
+	@Test
+	public void shouldGiveUserFiveSecretAnswerAttemptsAfterLessThanFiveFailedUsernameAttempts() throws Exception {
+		ForgotPasswordFormController controller = new ForgotPasswordFormController();
+		
+		for (int x = 1; x <= 4; x++) {
+			MockHttpServletRequest request = new MockHttpServletRequest();
+			request.setMethod("POST");
+			
+			request.addParameter("uname", "invaliduser");
+			
+			controller.handleRequest(request, new MockHttpServletResponse());
+		}
+		
+		// those were the first four, now the fifth is a valid username
+		MockHttpServletRequest request = new MockHttpServletRequest();
+		request.setMethod("POST");
+		
+		request.addParameter("uname", "validuser");
+		
+		controller.handleRequest(request, new MockHttpServletResponse());
+		
+		Assert.assertNotNull(request.getAttribute("secretQuestion"));
+		
+		// now the user has 5 chances at the secret answer
+		
+		// fifth request
+		MockHttpServletRequest request5 = new MockHttpServletRequest();
+		request5.setMethod("POST");
+		
+		request5.addParameter("uname", "validuser");
+		request5.addParameter("secretAnswer", "invalid answer");
+		controller.handleRequest(request5, new MockHttpServletResponse());
+		Assert.assertNotNull(request5.getAttribute("secretQuestion"));
+		
+		// sixth request (should not lock out because is after valid username)
+		MockHttpServletRequest request6 = new MockHttpServletRequest();
+		request6.setMethod("POST");
+		
+		request6.addParameter("uname", "validuser");
+		request6.addParameter("secretAnswer", "invalid answer");
+		request.setMethod("POST");
+		controller.handleRequest(request6, new MockHttpServletResponse());
+		Assert.assertNotNull(request6.getAttribute("secretQuestion"));
+		
+		// seventh request (should authenticate with valid answer)
+		MockHttpServletRequest request7 = new MockHttpServletRequest();
+		request7.setMethod("POST");
+		
+		request7.addParameter("uname", "validuser");
+		request7.addParameter("secretAnswer", "valid secret answer");
+		controller.handleRequest(request7, new MockHttpServletResponse());
+		
+		Assert.assertTrue(Context.isAuthenticated());
+	}
+	
+	@Test
+	public void shouldAuthenticateWithInValidSecretQuestionIfUserIsNull() throws Exception {
```

Comment:
This seems wrong to me, why would you authenticate a request with an invalod secret question and null user? This should fail